File: Daily/sbgo/Leetcode839-相似的字符串.go
package sbgo

type unionFind struct {
	parent, rank []int
	Area         int
}

func newUnionFind(n int) *unionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	Area := n
	for i := range parent {
		parent[i] = i
		rank[i] = 1
	}
	return &unionFind{parent: parent, rank: rank, Area: Area}
}

func (uf *unionFind) find(x int) int {
	if x == uf.parent[x] {
		return x
	}
	uf.parent[x] = uf.find(uf.parent[x])
	return uf.parent[x]
}

func (uf *unionFind) union(x, y int) {
	fx, fy := uf.find(x), uf.find(y)
	if fx == fy {
		return
	}
	if uf.rank[fx] < uf.rank[fy] {
		fx, fy = fy, fx
	}
	uf.parent[fy] = fx
	uf.rank[fx] += uf.rank[fy]
	uf.Area--
}

func numSimilarGroups(strs []string) int {
	n := len(strs)
	uf := newUnionFind(n)
	check := func(strs1, strs2 string) bool {
		num := 0
		for i := range strs1 {
			if strs1[i] != strs2[i] {
				num++
			}
			if num > 2 {
				return false
			}
		}
		return true
	}
	for i := 0; i < n-1; i++ {
		for j := 1; j < n; j++ {
			if strs[i] == strs[j] {
				uf.union(i, j)
				continue
			}
			if check(strs[i], strs[j]) {
				uf.union(i, j)
			}
		}
	}
	return uf.Area
}

File: Leetcode/PY/Leetcode839-相似的字符串.py
from typing import List
class UnionFind:
    def __init__(self,n):
        self.parent = list(range(n))
        self.rank   = [1 for _ in range(n)]
        self.Area = n
    def find(self,x):
        if x == self.parent[x]:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def union(self,x,y):
        fx,fy = self.find(x),self.find(y)
        if fx == fy:
            return 
        if self.rank[fx] < self.rank[fy]:
            fx,fy = fy,fx
        self.parent[fy] = fx
        self.rank[fx] += self.rank[fy]
        self.Area -= 1

  
class Solution:
    def numSimilarGroups(self, strs: List[str]) -> int:
        n = len(strs)
        uf = UnionFind(n)
        def check(strs1,strs2):
            num = 0
            ll = len(strs1)
            for i in range(ll):
                if strs1[i] != strs2[i]:
                    num += 1
                if num > 2:
                    return False
            return True
        for i in range(n-1):
            for j in range(i+1,n):
                if strs[i] == strs[j]:
                    uf.union(i,j)
                    break
                if check(strs[i],strs[j]):
                    uf.union(i,j)
        return uf.Area